tsnet: TestUDPConn flake, uploading health change

At 262c526c4efcbea0148bfade1fb0abfeea674aea + unrelated changes, looks like this test has a goroutine uploading its health change info to control after shutdown.

Maybe that async health upload goroutine needs to be attached to the context of the test shutting down. Or just disabled in tests.

```
=== RUN   TestUDPConn
    tsnet_test.go:116: DERP httpsrv listener: 127.0.0.1:55172
    tsnet_test.go:128: testcontrol listening on http://127.0.0.1:55174/
2024/10/02 15:00:29 tsnet starting with hostname "s1", varRoot "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestUDPConn2217018208\\001\\s1"
2024/10/02 15:00:30 LocalBackend state is NeedsLogin; running StartLoginInteractive...
2024/10/02 15:00:30 tsnet starting with hostname "s2", varRoot "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestUDPConn2217018208\\002\\s2"
2024/10/02 15:00:31 LocalBackend state is NeedsLogin; running StartLoginInteractive...
    tsnet_test.go:773: ping success: &ipnstate.PingResult{IP:"100.64.0.1", NodeIP:"100.64.0.1", NodeName:"s1", Err:"", LatencySeconds:0.005924, Endpoint:"", DERPRegionID:0, DERPRegionCode:"", PeerAPIPort:0x0, PeerAPIURL:"", IsLocalIP:false}
    tsnet_test.go:798: got: "hello"
    tsnet_test.go:818: got: "world"
    stuntest.go:65: STUN server shutdown
    resource.go:46: goroutine diff:
          bytes.Join({
        - 	"goroutine profile: total 4",
        - 	"1 @ 0x144551 0x18485d 0x312cf1 0x312b25 0x30f94b 0xcacd34 0xcac932 0xcde34b 0x267aeb 0x18e461",
        - 	`#	0x312cf0	runtime/pprof.writeRuntimeProfile+0xb0	C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:793`,
        - 	`#	0x312b24	runtime/pprof.writeGoroutine+0x44	C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:752`,
        - 	`#	0x30f94a	runtime/pprof.(*Profile).WriteTo+0x14a	C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:374`,
        - 	`#	0xcacd33	tailscale.com/tstest.goroutines+0x53	D:/a/tailscale/tailscale/tstest/resource.go:54`,
        - 	`#	0xcac931	tailscale.com/tstest.ResourceCheck+0x51	D:/a/tailscale/tailscale/tstest/resource.go:29`,
        - 	`#	0xcde34a	tailscale.com/tsnet.TestUDPConn+0x4a	D:/a/tailscale/tailscale/tsnet/tsnet_test.go:755`,
        - 	`#	0x267aea	testing.tRunner+0xca			C:/hostedtoolcache/windows/go/1.23.1/x64/src/testing/testing.go:1690`,
        + 	"goroutine profile: total 8",
        + 	"5 @ 0x185a4e 0x164012 0x163fe6 0x186d65 0x19f6c8 0x810c45 0x80c5d4 0x80d66e 0x806319 0x18e461",
        + 	`#	0x186d64	sync.runtime_Semacquire+0x24						C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/sema.go:71`,
        + 	`#	0x19f6c7	sync.(*WaitGroup).Wait+0x47						C:/hostedtoolcache/windows/go/1.23.1/x64/src/sync/waitgroup.go:118`,
        + 	`#	0x810c44	tailscale.com/util/singleflight.(*Group[...]).Do+0x1a4			D:/a/tailscale/tailscale/util/singleflight/singleflight.go:109`,
        + 	`#	0x80c5d3	tailscale.com/control/controlclient.(*NoiseClient).getConn+0x173	D:/a/tailscale/tailscale/control/controlclient/noise.go:232`,
        + 	`#	0x80d66d	tailscale.com/control/controlclient.(*NoiseClient).post+0x22d		D:/a/tailscale/tailscale/control/controlclient/noise.go:401`,
        + 	`#	0x806318	tailscale.com/control/controlclient.(*Direct).ReportHealthChange+0x1d8	D:/a/tailscale/tailscale/control/controlclient/direct.go:1633`,
        + 	"",
        + 	"1 @ 0x144551 0x18485d 0x312cf1 0x312b25 0x30f94b 0xcacd34 0xcacad7 0x266509 0x2670d5 0x267b85 0x267b15 0x18e461",
        + 	`#	0x312cf0	runtime/pprof.writeRuntimeProfile+0xb0		C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:793`,
        + 	`#	0x312b24	runtime/pprof.writeGoroutine+0x44		C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:752`,
        + 	`#	0x30f94a	runtime/pprof.(*Profile).WriteTo+0x14a		C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/pprof/pprof.go:374`,
        + 	`#	0xcacd33	tailscale.com/tstest.goroutines+0x53		D:/a/tailscale/tailscale/tstest/resource.go:54`,
        + 	`#	0xcacad6	tailscale.com/tstest.ResourceCheck.func1+0xd6	D:/a/tailscale/tailscale/tstest/resource.go:42`,
        + 	`#	0x266508	testing.(*common).Cleanup.func1+0x108		C:/hostedtoolcache/windows/go/1.23.1/x64/src/testing/testing.go:1176`,
        + 	`#	0x2670d4	testing.(*common).runCleanup+0xd4		C:/hostedtoolcache/windows/go/1.23.1/x64/src/testing/testing.go:1[354](https://github.com/tailscale/tailscale/actions/runs/11146255699/job/30977777645?pr=13644#step:5:355)`,
        + 	`#	0x267b84	testing.tRunner.func2+0x24			C:/hostedtoolcache/windows/go/1.23.1/x64/src/testing/testing.go:1684`,
        + 	`#	0x267b14	testing.tRunner+0xf4				C:/hostedtoolcache/windows/go/1.23.1/x64/src/testing/testing.go:1696`,
          	"",
          	"1 @ 0x185a4e 0x11a43e 0x119ff2 0x268912 0x26abd7 0x267aeb 0x26aabd 0x2694ea 0xce235b 0x150f5d 0x18e461",
          	... // 15 identical lines
          	`#	0x80f999	tailscale.com/util/singleflight.(*Group[...]).doCall+0x79		D:/a/tailscale/tailscale/util/singleflight/singleflight.go:297`,
          	`#	0x810bec	tailscale.com/util/singleflight.(*Group[...]).Do+0x14c			D:/a/tailscale/tailscale/util/singleflight/singleflight.go:123`,
        - 	`#	0x80c5d3	tailscale.com/control/controlclient.(*NoiseClient).getConn+0x173	D:/a/tailscale/tailscale/control/controlclient/noise.go:232`,
        - 	`#	0x80d66d	tailscale.com/control/controlclient.(*NoiseClient).post+0x22d		D:/a/tailscale/tailscale/control/controlclient/noise.go:401`,
        - 	`#	0x806318	tailscale.com/control/controlclient.(*Direct).ReportHealthChange+0x1d8	D:/a/tailscale/tailscale/control/controlclient/direct.go:1633`,
        - 	"",
        - 	"1 @ 0x185a4e 0x164012 0x163fe6 0x186d65 0x19f6c8 0x810c45 0x80c5d4 0x80d66e 0x806319 0x18e461",
        - 	`#	0x186d64	sync.runtime_Semacquire+0x24						C:/hostedtoolcache/windows/go/1.23.1/x64/src/runtime/sema.go:71`,
        - 	`#	0x19f6c7	sync.(*WaitGroup).Wait+0x47						C:/hostedtoolcache/windows/go/1.23.1/x64/src/sync/waitgroup.go:118`,
        - 	`#	0x810c44	tailscale.com/util/singleflight.(*Group[...]).Do+0x1a4			D:/a/tailscale/tailscale/util/singleflight/singleflight.go:109`,
          	... // 5 identical lines
          }, "\n")
        
    resource.go:47: goroutine count: expected 4, got 8
--- FAIL: TestUDPConn (5.93s)
FAIL	tailscale.com/tsnet
```

--- a/control/controlclient/direct.go
+++ b/control/controlclient/direct.go
@@ -82,6 +82,8 @@ type Direct struct {
 	onControlTime              func(time.Time)              // or nil
 	onTailnetDefaultAutoUpdate func(bool)                   // or nil
 	panicOnUse                 bool                         // if true, panic if client is used (for testing)
+	closedCtx                  context.Context              // alive until Direct.Close is called
+	closeCtx                   context.CancelFunc           // cancels closedCtx
 
 	dialPlan ControlDialPlanner // can be nil
 
@@ -303,6 +305,8 @@ func NewDirect(opts Options) (*Direct, error) {
 		dnsCache:                   dnsCache,
 		dialPlan:                   opts.DialPlan,
 	}
+	c.closedCtx, c.closeCtx = context.WithCancel(context.Background())
+
 	if opts.Hostinfo == nil {
 		c.SetHostinfo(hostinfo.New())
 	} else {
@@ -325,6 +329,8 @@ func NewDirect(opts Options) (*Direct, error) {
 
 // Close closes the underlying Noise connection(s).
 func (c *Direct) Close() error {
+	c.closeCtx()
+
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	if c.noiseClient != nil {
@@ -1628,7 +1634,7 @@ func (c *Direct) ReportHealthChange(w *health.Warnable, us *health.UnhealthyStat
 	}
 
 	// Best effort, no logging:
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	ctx, cancel := context.WithTimeout(c.closedCtx, 5*time.Second)
 	defer cancel()
 	res, err := np.post(ctx, "/machine/update-health", nodeKey, req)
 	if err != nil {

